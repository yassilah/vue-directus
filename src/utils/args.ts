import { Ref, computed, unref } from 'vue'
import { isFunction } from '@vue/shared'

export type ReactiveArg<T> = T | Ref<T> | (() => T)

/**
 * Use a reactive argument.
 */
export function reactiveArg<T>(value: ReactiveArg<T>) {
    return computed(() => {
        if (isFunction(value)) {
            return value()
        }

        return unref(value)
    })
}
