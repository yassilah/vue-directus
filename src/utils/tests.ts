import { App, createApp } from 'vue'
import install from '../index'

/**
 * Run a test with the plugin installed.
 */
export function useApp<T extends (...args: any[]) => any>(
    composable: T
): [ReturnType<T>, App<Element>]
export function useApp(): App<Element>
export function useApp<T extends (...args: any[]) => any>(composable?: T) {
    let result: ReturnType<T>

    const app = createApp({
        setup() {
            if (composable) {
                result = composable()
            }

            return () => null
        }
    })

    app.use(install)

    app.mount(document.createElement('div'))

    return composable ? [result, app] : app
}
