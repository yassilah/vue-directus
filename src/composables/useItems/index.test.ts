import { describe, expect, it } from 'vitest'
import { unref } from 'vue'
import { useApp } from '../../utils/tests'
import { useItems } from '.'

describe('useCollection', () => {
    it('add a composable', () => {
        const [result] = useApp(() => useItems('items'))

        expect(unref(result)).toEqual([])
    })

    it('should have controls', () => {
        const [result] = useApp(() =>
            useItems('items', {
                controls: true
            })
        )

        expect(unref(result)).toMatchSnapshot()
    })
})
