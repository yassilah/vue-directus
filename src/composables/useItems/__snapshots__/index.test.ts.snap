// Vitest Snapshot v1

exports[`useCollection > should have controls 1`] = `
{
  "createMany": [Function],
  "createOne": [Function],
  "deleteMany": [Function],
  "deleteOne": [Function],
  "errors": {
    "createMany": null,
    "createOne": null,
    "deleteMany": null,
    "deleteOne": null,
    "fetch": null,
    "updateMany": null,
    "updateOne": null,
  },
  "fetch": [Function],
  "items": RefImpl {
    "__v_isRef": true,
    "__v_isShallow": false,
    "_rawValue": [],
    "_value": [],
    "dep": undefined,
  },
  "loading": {
    "createMany": false,
    "createOne": false,
    "deleteMany": false,
    "deleteOne": false,
    "fetch": true,
    "updateMany": false,
    "updateOne": false,
  },
  "updateMany": [Function],
  "updateOne": [Function],
}
`;
